### PR TITLE
Combining dim red in one widget

### PR DESCRIPTION
**Pull request by Cryaaa**
I was trying to combine both algorithms into one widget so the plugin doesn't get too cluttered with all the different methods. It worked for me, so maybe you can try if it works for you? I think this would be the way to go if it works, but maybe others disagree?

### CODE REVIEW

File: napari_clusters_plotter/_dimension_reduction.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+import napari
+from qtpy.QtWidgets import QWidget, QPushButton, QLabel, QHBoxLayout, QVBoxLayout
+from qtpy.QtWidgets import QListWidget, QListWidgetItem, QAbstractItemView, QComboBox, QSpinBox
+from qtpy.QtCore import QRect
+from ._utilities import widgets_inactive
+from napari_tools_menu import  register_dock_widget
+
+@register_dock_widget(menu="Measurement > Dimensionality reduction (ncp)")
+
+class dimredWidget(QWidget):
```
Comment (haesleinhuepf) [inline napari_clusters_plotter/_dimension_reduction.py:12]:
```suggestion
class DimensionalityReductionWidget(QWidget):
```
Please use proper classnames. Also Python classes start upper case.

File: napari_clusters_plotter/_dock_widget.py [hunk 2]
```diff
@@ -2,9 +2,11 @@
...
 from ._umap import UMAPWidget
+from ._tsne import TSNEWidget
+from ._dimension_reduction import dimredWidget
 from ._kmeans_clustering import ClusteringWidget
 
 
 @napari_hook_implementation
 def napari_experimental_provide_dock_widget():
-    return [MeasureWidget, PlotterWidget, UMAPWidget, ClusteringWidget]
+    return [MeasureWidget, PlotterWidget, UMAPWidget, ClusteringWidget,TSNEWidget,dimredWidget]
```
Comment (haesleinhuepf) [inline napari_clusters_plotter/_dock_widget.py:12]:
```suggestion
    return [MeasureWidget, PlotterWidget, UMAPWidget, ClusteringWidget,TSNEWidget,DimensionalityReductionWidget]
```

File: napari_clusters_plotter/_dock_widget.py [hunk 3]
```diff
@@ -2,9 +2,11 @@
 from ._measure import MeasureWidget
 from ._plotter import PlotterWidget
 from ._umap import UMAPWidget
+from ._tsne import TSNEWidget
+from ._dimension_reduction import dimredWidget
```
Comment (haesleinhuepf) [inline napari_clusters_plotter/_dock_widget.py:6]:
```suggestion
from ._dimension_reduction import DimensionalityReductionWidget
```

File: napari_clusters_plotter/_dimension_reduction.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        self.label_list = QComboBox()
+        self.update_label_list()
+        self.label_list.currentIndexChanged.connect(self.update_properties_list)
+        choose_img_container.layout().addWidget(QLabel("Labels layer"))
+        choose_img_container.layout().addWidget(self.label_list)
+
+        # selection of dimension reduction algorithm
+        algorithm_container = QWidget()
+        algorithm_container.setLayout(QHBoxLayout())
+        algorithm_container.layout().addWidget(QLabel("Dimension Reduction Algorithm"))
```
Comment (haesleinhuepf) [inline napari_clusters_plotter/_dimension_reduction.py:41]:
```suggestion
        algorithm_container.layout().addWidget(QLabel("Dimensionality Reduction Algorithm"))
```

File: napari_clusters_plotter/_dimension_reduction.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+
+            # write result back to properties
+            for i in range(0, n_components):
+                properties["TSNE_" + str(i)] = embedding[:,i]
+
+        else:
+            warnings.warn('No Dimension Reduction Algorithm Chosen!')
+
+
+
```
Comment (haesleinhuepf) [inline napari_clusters_plotter/_dimension_reduction.py:201]:
```suggestion
```

File: napari_clusters_plotter/_tsne.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+        napari_viewer.layers.selection.events.changed.connect(self._on_selection)
+
+        self.current_annotation = None
+
+        # QVBoxLayout - lines up widgets vertically
+        self.setLayout(QVBoxLayout())
+        label_container = QWidget()
+        label_container.setLayout(QVBoxLayout())
+        label_container.layout().addWidget(QLabel("<b>Dimensionality reduction: TSNE</b>"))
```
Comment (haesleinhuepf) [inline napari_clusters_plotter/_tsne.py:26]:
```suggestion
        label_container.layout().addWidget(QLabel("<b>Dimensionality reduction: t-SNE</b>"))
```
I think the algorithm is spelled like that, right?

File: napari_clusters_plotter/_tsne.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
+import pandas as pd
+import warnings
+import napari
+from qtpy.QtWidgets import QWidget, QPushButton, QLabel, QHBoxLayout, QVBoxLayout
+from qtpy.QtWidgets import QListWidget, QListWidgetItem, QAbstractItemView, QComboBox, QSpinBox
+from qtpy.QtCore import QRect
+from napari_tools_menu import  register_dock_widget
+
+@register_dock_widget(menu="Measurement > TSNE dimensionality reduction (ncp)")
```
Comment (haesleinhuepf) [inline napari_clusters_plotter/_tsne.py:9]:
```suggestion
@register_dock_widget(menu="Measurement > t-SNE dimensionality reduction (ncp)")
```

File: napari_clusters_plotter/_dimension_reduction.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        self.label_list.currentIndexChanged.connect(self.update_properties_list)
+        choose_img_container.layout().addWidget(QLabel("Labels layer"))
+        choose_img_container.layout().addWidget(self.label_list)
+
+        # selection of dimension reduction algorithm
+        algorithm_container = QWidget()
+        algorithm_container.setLayout(QHBoxLayout())
+        algorithm_container.layout().addWidget(QLabel("Dimension Reduction Algorithm"))
+        self.algorithm_choice_list = QComboBox()
+        self.algorithm_choice_list.addItems(['   ', 'UMAP', 'TSNE'])
```
Comment (haesleinhuepf) [inline napari_clusters_plotter/_dimension_reduction.py:43]:
```suggestion
        self.algorithm_choice_list.addItems(['   ', 'UMAP', 't-SNE'])
```

**Review by haesleinhuepf — COMMENTED** [review]
Hi Ryan @Cryaaa ,

it looks very good to me. I also tried and it works like a charm. I added some suggestions to the code. Please check if it still works after considering my suggestions.

Furthermore, when running the TSNE, I had this warning popping up in napari:
> The PCA initialization in TSNE will change to have the standard deviation of PC1 equal to 1e-4 in 1.2. This will ensure better convergence.

Is there anything we can do about it?

**Comment by Cryaaa** [comment]
@haesleinhuepf Nice one! I'll test if the changes broke anything now and try and get rid of the warning.

**Comment by haesleinhuepf** [comment]
@Cryaaa: In the future, better leave merging to Laura @lazigu . She's the boss here :-)

**Comment by Cryaaa** [comment]
@haesleinhuepf @lazigu Oh, sorry I didn't realise it was going to merge into main. I'll definitely leave it to laura in the future... I thought I branched it off another branch I generated and that it would merge into that branch (If that makes any sense). I also looked at the error and it seems that the code in scikit learn spits out that message automatically if the initialisation is chosen as PCA, so until they remove the message it will always be displayed. This is slightly annoying as PCA is the initialisation method they suggest to use :D

**Comment by zoccoler** [comment]
> @haesleinhuepf @lazigu Oh, sorry I didn't realise it was going to merge into main. I'll definitely leave it to laura in the future... I thought I branched it off another branch I generated and that it would merge into that branch (If that makes any sense).

@Cryaaa It does to me! I had the exact same problem and reasoning you did. I was writing the new entry to the blog. I forked the repository, had to update my "local" main branch to solve differences before a PR and then I applied a merge thinking I was doing it locally, but it actually applied to the main repository...

> I also looked at the error and it seems that the code in scikit learn spits out that message automatically if the initialisation is chosen as PCA, so until they remove the message it will always be displayed. 

Did you try to remove all warnings before that? May be a little too extreme, but I would expect not to see that warning if you put this before that:

```
import warnings
warnings.filterwarnings("ignore")
```
It seems it is a `FutureWarning` [category](https://docs.python.org/3/library/warnings.html#warning-categories) as dislayed [here](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/manifold/_t_sne.py), so you could also potentially filter it by its type, similar to [this](https://docs.python.org/3/library/warnings.html#warning-categories).

**Comment by zoccoler** [comment]
Actually this seems to work:
`warnings.filterwarnings(action='ignore', category=FutureWarning, module='sklearn')`
Adapted from [here](https://stackoverflow.com/a/63980460/11885372)

**Comment by lazigu** [comment]
@Cryaaa no worries, I would have merged it anyway 😀 I really like this solution of having both options in the same widget. I will just change the file name to dimensionality_reduction and remove UMAP from the widget label.

Just for the future, it would super nice if you changed _umap.py directly instead of creating a new file, so not all file would appear green, it would be easier to review the code 😀

@zoccoler Thanks a lot! I will try this out.

**Comment by Cryaaa** [comment]
@lazigu actually the Widget is already called dimensionality reduction widget in the main branch. Additionally there are two widgets: one for umap and one for tsne. I thought we should leave them for now maybe but ultimately we could delete them. Hope that clears things up

**Comment by lazigu** [comment]
Oh sorry, I think I got confused because I just deleted them before reading your message 🙈 Why do you want to leave them if both methods are implemented in dim_red widget? It's a bit confusing if there are too many widgets, no?

**Comment by Cryaaa** [comment]
@lazigu yeah that makes sense, I just didn’t want to delete things without asking. I feel more comfortable just adding new files but I must admit it didn’t make sense 😄

**Comment by lazigu** [comment]
I just realized that I forgot to push my changes, but now I did that so you can see minor changes that I did.
Now I see what you meant, that to change a file name old file is deleted, my mistake, nevermind my comment about green code then 😀
Thanks a lot for this again 🤩
P.s. Marcelo's solution worked!
P.p.s @Cryaaa will you be working now on more contributions? Because I'm working now to simplify a lot of parts of the code using magicgui, so I can already see a lot of conflicts in the future if PRs keep coming 😂

**Comment by Cryaaa** [comment]
@lazigu ahhhhh I see that could indeed be a problem. I was going to work on a few other things but let’s talk about it tomorrow after the group meeting if you have some time?

**Comment by lazigu** [comment]
@Cryaaa yes, sure! Goodnight 😀
